test/integration/static_analysis/variables.spec.ts

```diff
@@ -0,0 +1,151 @@
+    const engine = new Liquid()
+    const template = engine.parse('{{ a }}')
+    const analysis = analyze(template)
+
+    const a = new Variable(['a'], { row: 1, col: 4, file: undefined })
+
+    expect(analysis).toStrictEqual({
+      variables: { a: [a] },
+      globals: { a: [a] },
+      locals: {}
+    })
+  })
+
+  it('should report all locations of a variable', () => {
+    const engine = new Liquid()
+    const template = engine.parse('{{ a }}\n{{ a }}')
+    const analysis = analyze(template)
+
+    const as = [
+      new Variable(['a'], { row: 1, col: 4, file: undefined }),
+      new Variable(['a'], { row: 2, col: 4, file: undefined })
+    ]
+
+    expect(analysis).toStrictEqual({
+      variables: { a: as },
+      globals: { a: as },
+      locals: {}
+    })
+  })
+
+  it('should report variables in filter arguments', () => {
+    const engine = new Liquid()
+    const template = engine.parse('{{ a | join: b }}')
+    const analysis = analyze(template)
+
+    const a = new Variable(['a'], { row: 1, col: 4, file: undefined })
+    const b = new Variable(['b'], { row: 1, col: 14, file: undefined })
+
+    expect(analysis).toStrictEqual({
+      variables: { a: [a], b: [b] },
+      globals: { a: [a], b: [b] },
+      locals: {}
+    })
+  })
+
+  it('should report dotted properties', () => {
+    const engine = new Liquid()
+    const template = engine.parse('{{ a.b }}')
+    const analysis = analyze(template)
+
+    const v = new Variable(['a', 'b'], { row: 1, col: 4, file: undefined })
+
+    expect(analysis).toStrictEqual({
+      variables: { 'a.b': [v] },
+      globals: { 'a.b': [v] },
+      locals: {}
+    })
+  })
+
+  it('should handle quoted properties', () => {
+    const engine = new Liquid()
+    const template = engine.parse('{{ a["b c"] }}')
+    const analysis = analyze(template)
+
+    const v = new Variable(['a', 'b c'], { row: 1, col: 4, file: undefined })
+
+    expect(analysis).toStrictEqual({
+      variables: { "a['b c']": [v] },
+      globals: { "a['b c']": [v] },
+      locals: {}
+    })
+  })
+
+  it('should report nested variables', () => {
+    const engine = new Liquid()
+    const template = engine.parse('{{ a[b.c] }}')
+    const analysis = analyze(template)
+
+    const bc = new Variable(['b', 'c'], { row: 1, col: 6, file: undefined })
+    const a = new Variable(['a', bc], { row: 1, col: 4, file: undefined })
+
+    expect(analysis).toStrictEqual({
+      variables: { 'a[b.c]': [a], 'b.c': [bc] },
+      globals: { 'a[b.c]': [a], 'b.c': [bc] },
+      locals: {}
+    })
+  })
+
+  it('should detect local variables', () => {
+    const engine = new Liquid()
+    const template = engine.parse('{% assign a = "foo" %}{{ a }}')
+    const analysis = analyze(template)
+
+    expect(analysis).toStrictEqual({
+      variables: { a: [new Variable(['a'], { row: 1, col: 26, file: undefined })] },
+      globals: { },
+      locals: { a: [new Variable(['a'], { row: 1, col: 1, file: undefined })] }
```

Comment:
I guess locals is for variables with a local scope, they can be local only (not depend on global variables), and also can be derived from global variables. In general people also want a list of access to their global variables. Like

```liquid
{% render 'product' with featured_product as product %}
// in "product.liquid"
{{product.foo}}
```

What we actually care about is `featured_product.foo`. How do you think for this scenario?

Reply:
I'll need to think about that. A similar situation could arise from a simple `{% assign %}` tag.

```liquid
{% assign product = featured_product %}
{{ product.foo }}
```
or even

```liquid
{% assign thing = featured_product.foo %}
{{ thing.bar }}
```

Reply:
There is a subtle difference: Assign can be used with filters which is considered as forming new values, while ‘as’ is just alias.
But I think it’s OK to not support such details, especially after we introduced ‘render’ to deprecate ’include’. That means template authors already have a chance to declare them explicitly.